src/dash-table/utils/actions.js

```diff
@@ -53,23 +53,31 @@ export const clearSelection = {
     selected_cells: []
 };
 
-export function editColumnName(column, columns, headerRowIndex) {
+export function changeColumnHeader(column, columns, headerRowIndex, mergeDuplicateHeaders, newColumnName) {
+    let cloneColumn = R.mergeRight({}, column);
+    const maxLength = getHeaderRows(columns);
+    if (typeof column.name === 'string' && maxLength > 1) {
+            const newColumnNames = Array(maxLength).fill(column.name);
+            cloneColumn = R.mergeRight(column, {name: newColumnNames});
+    }
+    const transformedColumns =  columns.slice(0);
+    const columnIndex = columns.findIndex(col => col.id === column.id);
+    transformedColumns[columnIndex] = cloneColumn;
```

Comment:
Similarly, because of how `R.set` works, it's unnecessary to clone `columns` if the affected column above is not modified.